contracts/examples/DXswapSwapRelayer.sol

```diff
@@ -0,0 +1,296 @@
+        } else {
+            require(IERC20(tokenA).balanceOf(address(this)) >= amountInTokenA, 'DXswapRelayer: INSUFFICIENT_TOKEN_A');
+        }
+        require(IERC20(tokenB).balanceOf(address(this)) >= amountInTokenB, 'DXswapRelayer: INSUFFICIENT_TOKEN_B');
+
+        address pair = _pair(tokenA, tokenB, factory);
+
+        require(pair != address(0), 'DXswapRelayer: INVALID_PAIR_ADDRESS');
+
+        orderIndex = _OrderIndex();
+        orders[orderIndex] = Order({
+            tokenIn: amountInTokenA > 0 ? tokenA : tokenB,
+            tokenOut: amountInTokenA > 0 ? tokenB : tokenA,
+            amountIn: amountInTokenA > 0 ? amountInTokenA : amountInTokenB,
+            priceTolerance: priceTolerance,
+            minReserveA: minReserveA,
+            minReserveB: minReserveB,
+            oraclePair: pair,
+            deadline: deadline,
+            maxWindowTime: maxWindowTime,
+            oracleId: 0,
+            factory: factory,
+            executed: false
+        });
+
+        (uint256 reserveA, uint256 reserveB, ) = IDXswapPair(pair).getReserves();
+
+        /* Create an oracle to calculate average price before swap */
+        uint256 windowTime = _consultOracleParameters(
+            amountInTokenA,
+            amountInTokenB,
+            reserveA,
+            reserveB,
+            maxWindowTime
+        );
+        orders[orderIndex].oracleId = oracleCreator.createOracle(windowTime, pair);
+        emit NewOrder(orderIndex);
+    }
+
+    function executeOrder(uint256 orderIndex) external {
+        Order storage order = orders[orderIndex];
+        require(orderIndex < orderCount, 'DXswapRelayer: INVALID_ORDER');
+        require(!order.executed, 'DXswapRelayer: ORDER_EXECUTED');
+        require(oracleCreator.isOracleFinalized(order.oracleId), 'DXswapRelayer: OBSERVATION_RUNNING');
+        require(block.timestamp <= order.deadline, 'DXswapRelayer: DEADLINE_REACHED');
+
+        address tokenIn = order.tokenIn;
+        uint256 amountOut;
+        amountOut = oracleCreator.consult(
+            order.oracleId,
+            tokenIn == address(0) ? IDXswapRouter(dxSwapRouter).WETH() : tokenIn,
+            order.amountIn
+        );
+
+        uint256 minAmountOut = amountOut.sub(amountOut.mul(order.priceTolerance) / PARTS_PER_MILLION);
+
+        order.executed = true;
+
+        _swap(tokenIn, order.tokenOut, order.amountIn, minAmountOut, order.factory);
+
+        emit ExecutedOrder(orderIndex);
+    }
+
+    // Updates a price oracle and sends a bounty to msg.sender
+    function updateOracle(uint256 orderIndex) external {
+        Order storage order = orders[orderIndex];
+        require(block.timestamp <= order.deadline, 'DXswapRelayer: DEADLINE_REACHED');
+        require(!oracleCreator.isOracleFinalized(order.oracleId), 'DXswapRelayer: OBSERVATION_ENDED');
+        uint256 amountBounty = GAS_ORACLE_UPDATE.mul(tx.gasprice).add(BOUNTY);
```

Comment:
I'd maybe cap the gas price to `BASEFEE` + a certain maximum amount as a miner tip (let's say 5 or 6 gwei?). I can see grief attacks where someone submits a tx with a high gas price, draining the contract of all ETH. At that point oracles become unworkable because there are no funds to pay bounties (it can also be bad if we want to perform an ETH trade because the ETH can be drained by a malicious user). The user would only gain the bounty, but it would be pretty damn bad for the DAO.

Reply:
Nice catch! Agree that would be good to use basefee, but to do that we need to upgrade Solidity, because `block.basefee` was introduced with 0.8.7. I suggest upgrading to the newest version anyways. 
I'm not sure about a miner tip, what do you think @jpkcambridge?